### ISSUE TITLE
Create an automated way of keeping track of whether sparsity pattern should be preserved or not

### ISSUE THREAD

**Issue by GiudGiud** [idaholab/moose#28814]
### Motivation

See https://github.com/idaholab/moose/pull/28736/files

### Design

Keep track of what the user mentioned for the Problem parameter
```
  error_on_jacobian_nonzero_reallocation = false
```

### Impact

Less exposure to sparsity pattern issues for the user

**Comment by lindsayad** [comment]
I don't understand this issue. Are you talking about possibly calling `PetscMatrix::reset_preallocation` if we are worried that later Jacobian assemblies may have different nonzero structure than earlier ones, and in which both sets of nonzeroes are subsets of the sparsity pattern we determine with our relationship managers?

**Comment by GiudGiud** [comment]
No I m talking about having those "0 * ad number" operations to preserve the sparsity pattern only for solvers that require it.

For example the segregated one does not care (I think), but the Newton solver does

**Comment by lindsayad** [comment]
So do conditional addition of those artificial terms for preserving sparsity based on the solver type? That seems like a fine idea

**Comment by GiudGiud** [comment]
Yes. But tie it to the actual need (sparsity pattern preservation) than pointing out specific solvers.

**Comment by grmnptr** [comment]
I don't think we need a generalized system for this. This is a problem for k-epsilon alone because there is a lot of limiting that discards derivatives. I still feel like we could fix this within the functor material and the two kernels alone. 
1. with not doing the early return and just overriding harmonic for that face to average if needed. That would automatically bring in the derivatives with respect to u no matter what. 
2. Then we could add the derivatives in the functor material somehow. We can check the k value, the epsilon value and just multiply their derivatives with 0 and add them to the muts derivatives even when we are limiting mut.
I wanted to try these today but couldn't get to it. I am more or less convinced that these would work and would result in a clean computeQp function.

**Comment by GiudGiud** [comment]
Again, this code:
```
if (coef = 0)
  do average
else
  do what the user requested
```
does not look any better than what we need to have now. 

The PR is a patch that was made the evening of the issue being reported. Now we are ten days later, you do not need to get to it, I ll do it when I am done with the conference.

**Comment by grmnptr** [comment]
I think it does look better because it says to the user: -hey if you have 0 coefficients don't do harmonic- instead of -if you have 0 coefficients create something here multiply it with zero and add that to make sure you keep a sparsity pattern that you have no idea about-. But if you think there is a way to do this nicer, I am all in for it. I just feel like we are doing all of this patch because of the shortcomings of k-epsilon in Newton, which is a very specific thing in a very specific module.
